pkg/services/auth/auth.go

```diff
@@ -0,0 +1,242 @@
+		a.gitClient = git.New(pubKey)
+
+		return nil
+	}
+
+	deployKey, secret, err := a.generateDeployKey(targetName, name, repo)
+	if err != nil {
+		return fmt.Errorf("error generating deploy key: %w", err)
+	}
+
+	publicKeyBytes := extractPublicKey(secret)
+
+	if err := a.gitProvider.UploadDeployKey(owner, repoName, publicKeyBytes); err != nil {
+		return fmt.Errorf("error uploading deploy key: %w", err)
+	}
+
+	if err := a.storeDeployKey(ctx, secret); err != nil {
+		return fmt.Errorf("error storing deploy key: %w", err)
+	}
+
+	a.logger.Println("Deploy key generated and uploaded to git provider")
+	// Set the gitClient to use the newly generated deploy key
+	a.gitClient = git.New(deployKey)
+
+	return nil
+
+}
+
+var ErrNoDeployKeysSetup = errors.New("the git client has not been intialized. Run SetupDeployKeys() on this object first.")
+
+// GitClient returns an instrumented git client that will use the deploy key any git operations.
+// SetupDeployKeys MUST be called before this, else an error will be returned.
+func (a *AuthService) GitClient() (git.Git, error) {
+	if a.gitClient == nil {
+		return nil, ErrNoDeployKeysSetup
+	}
+
+	return a.gitClient, nil
+}
+
+// GitProvider returns a GitProvider that is already constructed with access tokens.
+// This can be used for GitProvider API operations, like opening pull requests or adding deploy keys.
+func (a *AuthService) GitProvider() gitproviders.GitProvider {
+	return a.gitProvider
+}
+
+// Generates an ssh keypair for upload to the Git Provider and for use in a git.Git client.
+func (a *AuthService) generateDeployKey(targetName string, secretName types.NamespacedName, repo gitproviders.NormalizedRepoURL) (*ssh.PublicKeys, *corev1.Secret, error) {
+	secret, err := a.createKeyPairSecret(targetName, secretName, repo)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not create key-pair secret: %w", err)
+	}
+
+	privKeyBytes := extractPrivateKey(secret)
+
+	deployKey, err := makePublicKey(privKeyBytes)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error creating new public keys: %w", err)
+	}
+
+	return deployKey, secret, nil
+
+}
+
+// Wrapper to abstract how the key is stored in case we want to change this later.
+func (a *AuthService) storeDeployKey(ctx context.Context, secret *corev1.Secret) error {
+	if err := a.k8sClient.Create(ctx, secret); err != nil {
+		return fmt.Errorf("could not store secret: %w", err)
+	}
+
+	return nil
+}
+
+// Wrapper to abstract how a key is fetched.
+func (a *AuthService) retreiveDeployKey(ctx context.Context, name types.NamespacedName) (*corev1.Secret, error) {
+	secret := &corev1.Secret{}
+	if err := a.k8sClient.Get(ctx, name, secret); err != nil {
+		return nil, fmt.Errorf("error getting deploy key secret: %w", err)
+	}
+
+	return secret, nil
+}
+
+// Uses flux to create a ssh key pair secret.
+func (a *AuthService) createKeyPairSecret(targetName string, name types.NamespacedName, repo gitproviders.NormalizedRepoURL) (*corev1.Secret, error) {
```

Comment:
I really advise returning just the struct `corev1.Secret` instead of the pointer.  I can tell you from experience that copying is cheap in Go, it will make the code simpler and easier to reason about over the long term as well.

Reply:
I'm following how the official k8s client libs do it: https://pkg.go.dev/k8s.io/client-go@v0.22.0/kubernetes/typed/apps/v1#DeploymentInterface

You are going to have to pry my "always pass by reference" JavaScript habits from my cold, dead fingers :wink:. 

Let's come up with some sort of guide on this for the team?

Reply:
My last place of employment did the same to me.  I'm just passing on the habit :)

We were allocating a ridiculous number of structs and learned the hard way that pointers on the heap make machines sad if you allocate enough of them.  I get we won't meet the volume that we experienced.  I also find it easier to reason about.  Anyway, your reasoning for following the k8s standards is extremely valid as well.